File: play.py
from globals import *

__all__ = ['play_level']

def play_level():
    count = 0
    while True:
        g.interrupted = 0
        if g.hit_message:
            message(g.hit_message, 0)
            g.hit_message = ""
            
        move(rogue.row, rogue.col)
        refresh()
        
        ch = getchar()
        check_message()
        
        while True: # for "goto CH"
            if ch == '.':
                rest(count if count > 0 else 1)
            elif ch == 'i':
                inventory(rogue.pack, IS_OBJECT)
            #elif ch == 'p':
            #    inventory(g.level_objects, IS_OBJECT)
            elif ch == 'f':
                fight(0)
            elif ch == 'F':
                fight(1)
            elif ch in 'hjklyunb':
                single_move_rogue(ch, 1)
            elif ch in 'HJKLYUNB\010\012\013\014\031\025\016\002':
                multiple_move_rogue(ch)
            elif ch == 'e':
                eat()
            elif ch == 'q':
                quaff()
            elif ch == 'r':
                read_scroll()
            elif ch == 'm':
                move_onto()
            elif ch == 'd':
                drop()
            elif ch == '\020':
                remessage()
            elif ch == '>':
                if check_down():
                    return
            elif ch == '<':
                if check_up():
                    return
            elif ch == 'I':
                single_inventory()
            elif ch == '\022':
                wrefresh(curscr)
            elif ch == 'T':
                take_off()
            elif ch == 'W' or ch == 'P':
                wear()
            elif ch == 'w':
                wield()
            elif ch == 'c':
                call_it()
            elif ch == 'z':
                zapp()
            elif ch == 't':
                throw()
            elif ch == '\032':
                tstp()
            elif ch == '!':
                shell()
            elif ch == 'v':
                message("pyrogue: Version 1.0 (sma was here)", 0)
            elif ch == 'Q':
                quit()
            elif ch in "0123456789":
                count = 0
                while True:
                    count = 10 * count + ord(ch) - ord('0')
                    ch = getchar()
                    if ch < '0' or ch > '9': break
                continue # goto CH
            elif ch == ' ':
                pass
            else:
                message("unknown command")
            break

File: flrogue/lib/rogue/play.dart
import 'dart:async';

import 'globals.dart';
import 'hit.dart';
import 'inventory.dart';
import 'level.dart';
import 'message.dart';
import 'move.dart';
import 'pack.dart';
import 'room.dart';
import 'score.dart';
import 'throw.dart';
import 'ui.dart';
import 'use.dart';
import 'zap.dart';

Future<void> playLevel() async {
  int count = 0;

  while (true) {
    interrupted = false;
    if (hitMessage.isNotEmpty) {
      await message(hitMessage);
      hitMessage = "";
    }

    ui.move(rogue.row, rogue.col).refresh();

    String ch = await ui.getchar();
    checkMessage();

    while (true) {
      // for "goto CH"
      if (ch == '.') {
        await rest(count > 0 ? count : 1);
      } else if (ch == 'i') {
        await inventory(rogue.pack, Cell.isObject);
      } else if (ch == 'f') {
        await fight(false);
      } else if (ch == 'F') {
        await fight(true);
      } else if ('hjklyunb'.contains(ch)) {
        await singleMoveRogue(ch, true);
      } else if ('HJKLYUNB\x08\x0a\x0b\x0c\x19\x15\x0e\x02'.contains(ch)) {
        await multipleMoveRogue(ch);
      } else if (ch == 'e') {
        await eat();
      } else if (ch == 'q') {
        await quaff();
      } else if (ch == 'r') {
        await readScroll();
      } else if (ch == 'm') {
        await moveOnto();
      } else if (ch == 'd') {
        await drop();
      } else if (ch == '>') {
        if (await checkDown()) {
          return;
        }
      } else if (ch == '<') {
        if (await checkUp()) {
          return;
        }
      } else if (ch == 'I') {
        await singleInventory();
      } else if (ch == '\x12') {
        ui.refresh();
      } else if (ch == 'T') {
        await takeOff();
      } else if (ch == 'W' || ch == 'P') {
        await wear();
      } else if (ch == 'w') {
        await wield();
      } else if (ch == 'c') {
        await callIt();
      } else if (ch == 'z') {
        await zapp();
      } else if (ch == 't') {
        await throwItem();
      } else if (ch == '\x17') {
        drawMagicMap();
      } else if (ch == 'v') {
        await message("pyrogue: Version 1.0 (dart port)");
      } else if (ch == 'Q') {
        await quit();
      } else if ('0123456789'.contains(ch)) {
        count = 0;
        while (true) {
          count = 10 * count + (ch.ascii - '0'.ascii);
          ch = await ui.getchar();
          if (!('0123456789'.contains(ch))) break;
        }
        continue; // goto CH
      } else if (ch == ' ') {
        // Do nothing
      } else {
        await message("unknown command");
      }
      break;
    }
  }
}
